task/golangci-lint/0.1/golangci-lint.yaml

```diff
@@ -0,0 +1,43 @@
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: golangci-lint
+  labels:
+    app.kubernetes.io/version: "0.1"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.36.0"
+    tekton.dev/categories: Code Quality
+    tekton.dev/tags: linter
+    tekton.dev/displayName: "golangci-lint"
+    tekton.dev/platforms: "linux/amd64"
+spec:
+  description: >-
+    The golangci-lint Task runs a golang linter.
+    Task is tailored to be triggered by prow as a prowjob.
+    It uses a golangci-lint tool.
+    The golangci tool config is stored in the .golangci.yaml file which is read from a repository hosting tested code.
+    A workspace named repo stores the source code to lint.
+  workspaces:
+    - name: repo
+      description: "The workspace where sources to lint are stored. Usually it should be a shared workspace with other tasks. A default mount path is aligned with prowjob running under a kubernetes agent."
+      mountPath: /home/prow/go
+  params:
+    - name: REPO_OWNER
+      description: "GitHub org that triggered the job. Variable set by prow. Prow will not set this variable if the prowjob type is periodic. In that case an empty string default value will be used."
+      type: string
+      default: ""
+    - name: REPO_NAME
+      description: "GitHub repo that triggered the job. Variable set by prow. Prow will not set this variable if the prowjob type is periodic. In that case an empty string default value will be used."
+      type: string
+      default: ""
+  steps:
+    - name: lint-go-code
+      image: eu.gcr.io/kyma-project/test-infra/golangci-lint:v20230221-85b2d35f
+      workingDir: /home/prow/go/src/github.com/$(params.REPO_OWNER)/$(params.REPO_NAME)
+      timeout: 30m
+      resources:
+        requests:
+          memory: 1.5Gi
+          cpu: 2
+      command:
+        - /golangci-lint.sh
```

Comment:
What would be the use-case here? This script was made generally to work around monorepos in Prow, so you don't need to define separate jobs for each module. And AFAIK you'll be using this job against single module. I believe calling golangci-lint directly would be safer bet. Additionally I believe you also should add flag overriding, as this task will run against *entire* codebase, not changes only for example.

Reply:
A use case is to run linter. A linter which is run is a golangci-lint. Exactly the same you added to the test-infra repository.

This task is a copy of existing solution from test-infra which I believe you created. Please provide some suggestions how to improve it.

Reply:
Will this work with test-infra then? Do we need two different tasks for test-infra and other repositories?

Reply:
Probably if you exclude the separate modules and add them as separate tasks, then yes it should. Right now ProwJob with script is working just fine for test-infra, but we call golangci-lint directly, bypassing this script for better customizability.

Reply:
Where we call golangci-lint directly? In test-infra? I saw our prowjob call a script not a tool directly.
Calling a script on repo with one module will not work?

Reply:
It's not hard to look at the repository ProwJobs. Here's some examples:
https://github.com/kyma-project/test-infra/blob/d0425460727fd7fc8e7bd0e250f487c0be870884/templates/data/lifecycle-manager-data.yaml#L21
https://github.com/kyma-project/test-infra/blob/3c1daf75679a5576379662cdbd532a30aa8696ff/templates/data/api-gateway-validation.yaml#L39

There is only *one* execution of this script and it's explicitly for the codebase of test-infra...
https://github.com/kyma-project/test-infra/blob/d0425460727fd7fc8e7bd0e250f487c0be870884/prow/jobs/test-infra/go-validation.yaml#L55

Reply:
It's not hard to be specific in comments and provide some links and details.
It's not hard to answer all questions I've asked.

I will ask third time. Hope this time it will not be to hard for you to answer a question.

Will golangci-lint.sh work correctly with repository with one module?

Reply:
Can't guarantee. We didn't test it though it should work fine, but with limited functionality as there is no way to pass additional flags to this wrapper. See https://github.com/kyma-project/test-infra/blob/main/prow/images/golangci-lint/golangci-lint.sh

Reply:
All right. Thanks for comments.
I will run test against a repo with single module to check how it will behave.

In terms of passing additional golangci-lint command line arguments I was thinking about it. I think it will be possible to do it with array params or ordinary string param, but I need to read some docs first. I don't have required knowledge yet about all available params types.

I will update Task description to make it clear it's mainly dedicated to run against monorepos and will not provide further features in this PR. Once I will know test run results and get knowledge about params I will think about improvements. This will be covered by separate issue.

Reply:
I've run a golangci-lint task against repo with one module. Task was able to run linter without problems.

```shell
Starting Scan
/home/prow/go/src/github.com/kyma-project/keda-manager
- Linting .
level=info msg="[config_reader] Config search paths: [./ /home/prow/go/src/github.com/kyma-project/keda-manager /home/prow/go/src/github.com/kyma-project /home/prow/go/src/github.com /home/prow/go/src /home/prow/go /home/prow /home / /root]"
level=info msg="[config_reader] Used config file .golangci.yaml"
level=info msg="[lintersdb] Active 7 linters: [errcheck gosimple govet ineffassign staticcheck typecheck unused]"
level=info msg="[loader] Go packages loading at mode 575 (imports|types_sizes|deps|exports_file|name|compiled_files|files) took 1m54.927332056s"
level=info msg="[runner/filename_unadjuster] Pre-built 0 adjustments in 10.60582ms"
level=info msg="[linters_context/goanalysis] analyzers took 5m5.745570442s with top 10 stages: buildir: 4m9.134174888s, nilness: 9.939761759s, fact_deprecated: 7.968739629s, inspect: 7.036695539s, ctrlflow: 6.956360622s, printf: 6.88834138s, fact_purity: 6.45554193s, typedness: 4.28275386s, SA5012: 3.674337702s, S1039: 153.03219ms"
level=info msg="[runner] processing took 64.15µs with stages: filename_unadjuster: 59.75µs, skip_dirs: 1.16µs, max_same_issues: 710ns, nolint: 630ns, max_from_linter: 240ns, cgo: 210ns, uniq_by_line: 200ns, max_per_file_from_linter: 180ns, skip_files: 170ns, autogenerated_exclude: 160ns, path_prettifier: 90ns, diff: 80ns, exclude-rules: 80ns, exclude: 80ns, identifier_marker: 80ns, sort_results: 80ns, source_code: 70ns, severity-rules: 60ns, path_shortener: 60ns, path_prefixer: 60ns"
level=info msg="[runner] linters took 51.309143195s with stages: goanalysis_metalinter: 51.308649715s"
level=info msg="File cache stats: 0 entries of total size 0B"
level=info msg="Memory: 1579 samples, avg is 311.2MB, max is 1487.9MB"
level=info msg="Execution took 2m46.259963389s"
- Linting ./test
level=info msg="[config_reader] Config search paths: [./ /home/prow/go/src/github.com/kyma-project/keda-manager/test /home/prow/go/src/github.com/kyma-project/keda-manager /home/prow/go/src/github.com/kyma-project /home/prow/go/src/github.com /home/prow/go/src /home/prow/go /home/prow /home / /root]"
level=info msg="[config_reader] Used config file ../.golangci.yaml"
level=info msg="[lintersdb] Active 7 linters: [errcheck gosimple govet ineffassign staticcheck typecheck unused]"
level=info msg="[loader] Go packages loading at mode 575 (compiled_files|exports_file|imports|name|types_sizes|deps|files) took 30.103331623s"
level=info msg="[runner/filename_unadjuster] Pre-built 0 adjustments in 1.108069ms"
level=info msg="[linters_context/goanalysis] analyzers took 1m0.86378332s with top 10 stages: buildir: 50.754822597s, printf: 1.705902158s, inspect: 1.481344106s, nilness: 1.417853626s, ctrlflow: 1.381256638s, fact_purity: 1.197628576s, fact_deprecated: 1.177107242s, SA5012: 1.009394403s, typedness: 673.261852ms, SA4009: 10.533339ms"
level=info msg="[runner] processing took 5.26µs with stages: skip_dirs: 1.14µs, max_same_issues: 680ns, nolint: 560ns, cgo: 370ns, autogenerated_exclude: 310ns, max_from_linter: 260ns, path_prettifier: 210ns, filename_unadjuster: 210ns, exclude-rules: 200ns, exclude: 190ns, identifier_marker: 190ns, max_per_file_from_linter: 180ns, skip_files: 180ns, source_code: 160ns, sort_results: 80ns, path_prefixer: 70ns, severity-rules: 70ns, path_shortener: 70ns, uniq_by_line: 70ns, diff: 60ns"
level=info msg="[runner] linters took 17.008632331s with stages: goanalysis_metalinter: 17.008388991s"
level=info msg="File cache stats: 0 entries of total size 0B"
level=info msg="Memory: 466 samples, avg is 203.9MB, max is 752.0MB"
level=info msg="Execution took 47.129192122s"
Scanning Finished

Step completed successfully
```